mathics/algorithm/parts.py

```diff
@@ -21,78 +24,99 @@
 
 SymbolNothing = Symbol("Nothing")
 
-# TODO: delete me
-# def join_lists(lists):
-#    """
-#    flatten a list of list.
-#    Maybe there are better, standard options, like
-#    https://stackoverflow.com/questions/952914/how-to-make-a-flat-list-out-of-a-list-of-lists.
-#    In any case, is not used in the following code.
-#    """
-#    new_list = []
-#    for list in lists:
-#        new_list.extend(list)
-#    return new_list
-
-
-def get_part(varlist, indices):
-    "Simple part extraction. indices must be a list of python integers."
-
-    def rec(cur, rest):
-        if rest:
-            if isinstance(cur, Atom):
-                raise PartDepthError(rest[0])
-            pos = rest[0]
-            elements = cur.elements
-            try:
-                if pos > 0:
-                    part = elements[pos - 1]
-                elif pos == 0:
-                    part = cur.get_head()
-                else:
-                    part = elements[pos]
-            except IndexError:
-                raise PartRangeError
-            return rec(part, rest[1:])
-        else:
-            return cur
 
-    return rec(varlist, indices).copy()
+def get_part(expression: BaseElement, indices: List[int]) -> BaseElement:
+    """Extract part of ``expression`` specified by ``indicies`` and
+    return that.
+    """
+
+    def get_subpart(expression: BaseElement, indices: List[int]) -> BaseElement:
+        """Recursive work-horse portion of ``get_part()`` that extracts pieces
+        of ``expression`` based on ``indices``. ``expression`` and ``indices``
+        are smaller parts of the corresponding variables at the outer level.
+        """
+        if not indices:
+            return expression
+
+        if isinstance(expression, Atom):
+            raise PartDepthError(indices[0])
+
+        pos = indices[0]
+        elements = expression.elements
+        try:
+            if pos > 0:
+                part = elements[pos - 1]
+            elif pos == 0:
+                part = expression.get_head()
+            else:
+                part = elements[pos]
+        except IndexError:
+            raise PartRangeError
+        return get_subpart(part, indices[1:])
 
+    return get_subpart(expression, indices).copy()
 
-def set_part(varlist, indices, newval):
-    "Simple part replacement. indices must be a list of python integers."
 
-    def rec(cur, rest):
-        if len(rest) > 1:
-            pos = rest[0]
-            if isinstance(cur, Atom):
+def set_part(expression, indices: List[int], newval) -> BaseElement:
+    """Replace all parts of ``expression`` specified by ``indicies`` with the
+    ``newval`` value. Return the modified compound expression.
+    """
+
+    def set_subpart(expression, indices: List[int]) -> BaseElement:
```

Comment:
in order to make this super-clear: maybe the parameters of `set_subpart` should be `sub_expression` and  `remaining_indices`.

Reply:
Ok, will do on next commit. And I suspect pylint (which in my opinion is by default too noisy) will like this better. I think it doesn't like hiding names this way in nested functions. (I don't totally agree with this, but I don't strongly object either.)

Reply:
Done via 65269755